hyperspy/hyperspy
# Split IO #2

### Description of the change

Split HyperSpy's IO plugins into a separate package RosettaSciIO. See #1978 for the related discussion.

This supersedes #2174

### Progress of the PR
- [ ] Change implemented (can be split into several points)
    - [x] Move all plugins to a separate folder, with its own setup.py, that must be installed separately.
    - [x] Move each plugin to a folder of its own
    - [x] Move all IO tests to rsciio
    - [x] Enable lazy loading of the plugins by moving the specifications to yamls files.
    - [x] Make rosettasciio fully independent of HyperSpy
        - [x] Don't use ``DictionaryTreeBrowser``. DTB was replaced by python-box
        - [x] Move ``misc.io`` code to ``rsciio``
        - [x] Move exceptions to ``rsciio``. 
        - [x] Move remaining imports to ``rsciio``Most of it is done, except for:
            - [x] ``rgb_tools``
            - [x]  ``emd`` relies on hyperspy for reading
        - [x] Make `_hierarchical` fully independent of HyperSpy
        - [x] Pass dictionary instead of signal to ``file_writer``s
            - [x] msa
            - [x] tvips
            - [x] semper
            - [x] blockfile
            - [x] emd
            - [x] panta rhei
            - [x] nexus
            - [x] image (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] mrcz (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] ripple (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] nexus (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] tiff (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] usid (https://github.com/francisco-dlp/hyperspy/pull/51)
            - [x] zspy
            - [x] hspy
    - [x] Move User Guide IO chapter to rsciio
    - [x] Remove dependency on traits (basically for t.Undefined)
    - [ ]  Make tests independent of hyperspy (not strictly required for release)
- [ ] update docstring (if appropriate),
- [x] update user guide (if appropriate),
- [x] add an changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/hyperspy/blob/RELEASE_next_minor/upcoming_changes/README.rst)) -- see #2979, 
- [ ] Check formatting changelog entry in the `readthedocs` doc build of this PR (link in github checks)

### How to test it

To test it, simply install `rosettasciio` (currently at hyperspy's root folder, it'll move to its own repository once there are no more code exchanges between hyperspy and RosettaSciIO) in addition to hyperspy. Everything should work as usual.

Additionally, it is possible to import the readers directly from RosettaSciIO as follows:

```python
from rsciio.msa import api
api.file_reader("your_msa_file.msa")
```
